Draft-Accepted/RFCXXXX-Command-Shell-Integration.md

```diff
@@ -0,0 +1,326 @@
+
+- `{commandLine}` - The full command line that the user has typed so far.
+- `{cursorPosition}` - The current cursor position within the command line, starting at 0 index.
+
+```json
+{
+    "$schema": "https://schemas.microsoft.com/commandshell/2023/05/01/command.shellintegration.json",
+    "executable": "az",
+    "version": "1.0.0",
+    "tabCompletion": {
+        "command": {
+            "executable": "az",
+            "arguments": [
+                "completion",
+                "powershell",
+                "--commandLine",
+                "{commandLine}",
+                "--cursorPosition",
+                "{cursorPosition}"
+            ]
+        }
+    }
+}
+```
+
+or a shell specific script:
+
+```json
+{
+    "$schema": "https://schemas.microsoft.com/commandshell/2023/05/01/command.shellintegration.json",
+    "executable": "az",
+    "version": "1.0.0",
+    "tabCompletion": {
+        "script": {
+            "powershell": "az-completion.ps1",
+            "bash": "az-completion.sh",
+            "zsh": "az-completion.zsh"
+        }
+    }
+}
+```
+
+Multiple shells can be specified and up to the shell to determine which one to use.
+Both a `command` and `script` can be specified, but only one of each and also up to the shell to determine which one to use.
+
+In this PowerShell example, the `az-completion.ps1` would be called one time to register the argument completer.
+A different shell may decide to call the script for each tab completion request.
+
+### Help Registration
+
+The same manifest may define a path to the help content:
+
+```json
+{
+    "$schema": "https://schemas.microsoft.com/commandshell/2023/05/01/command.shellintegration.json",
+    "executable": "az",
+    "version": "1.0.0",
+    "help": {
+        "tags": [
+            "azure",
+            "az"
+        ],
+        "path": "./help/az-help.md"
+    }
+}
+```
+
+Optional `tags` can be specified to help with searching for help content.
+
+### Predictors and Feedback Providers Registration
+
+Specifically for PowerShell, a command may also want to optionally register a Predictor or Feedback Provider.
+
+```json
+{
+    "$schema": "https://schemas.microsoft.com/commandshell/2023/05/01/command.shellintegration.json",
+    "executable": "az",
+    "version": "1.0.0",
+    "predictor": {
+        "powershell": "az-predictor.dll"
+    },
+    "feedbackProvider": {
+        "powershell": "az-feedbackprovider.dll"
+    }
+}
+```
+
+Similar to `tabCompletion`, alternate shells can choose to adopt their own mechanism for registering these extensions
+under their shell specific key.
+
+In the case of PowerShell, predictors and feedback providers are assemblies so the value is the path to the assembly
+instead of a `.ps1` script.
+To enable support for MSIX installed tools and not load the dll directly from their location, PowerShell will
+copy the dll to the user's folder if it is not already there (SHA256 hash comparison) and load it from there:
```

Comment:
A couple things:
1. The copying behavior probably should only be applied to MSIX native tools. For other tools, it may not be a problem to load in place. However, `Get-Module` will should the dll name, which is less friendly to the user.
2. When copying is needed, we need to consider dependency dlls too.

Maybe the manifest should point to a module folder instead of a list of dlls.

Reply:
A folder seems like a good solution here and the entire contents would be cached for MSIX.  For non-MSIX installed tools, they would still be blocked from updating (like MSI via WinGet) if we don't copy.

Reply:
We may run into inconsistencies in either way.
Imagine we have the tool A installed, and a PowerShell session is open and already have the predictor/completer/feedback-provider loaded in.

1. If we directly load from the installation location, then updating the tool may fail because those assemblies cannot be removed, and the user needs to close the PowerShell session to continue the update -- hopefully the MSI installer will tell the user what applications to close to continue.

2. If we copy and load, then the tool will be successfully updated, but then the user may find that the predictor/completer/feedback-provider stop working correctly because the tool was updated, but the loaded predictor/completer/feedback-provider is still the old version.

For an end user in the 2nd scenario, it would be mysterious to understand what is going wrong, and just have to try closing the current PowerShell session and start a new one.
I don't have a strong opinion which one a better user experience. I personally tend to the 1st one.

Reply:
The direct loading for MSIX is counter to the contract set for AppX, so we simply can't do that.  We'll have to see in the real world how often users hit this problem.